ASMDocumentation.md

```diff
@@ -0,0 +1,5841 @@
+         divsd   xmm0, xmm1
+```
+
+</details>
+
+<details>
+<summary>CPUDetectClass::Init_CPUID_Instruction</summary>
+
+```c++
+void CPUDetectClass::Init_CPUID_Instruction()
+{
+	unsigned long cpuid_available=0;
+
+   // The pushfd/popfd commands are done using emits
+   // because CodeWarrior seems to have problems with
+   // the command (huh?)
+
+#ifdef WIN32
+   __asm
+   {
+      mov cpuid_available, 0	// clear flag
+      push ebx
+      pushfd
+      pop eax
+      mov ebx, eax
+      xor eax, 0x00200000
+      push eax
+      popfd
+      pushfd
+      pop eax
+      xor eax, ebx
+      je done
+      mov cpuid_available, 1
+   done:
+      push ebx
+      popfd
+      pop ebx
+   }
+#elif defined(_UNIX)
+     __asm__(" mov $0, __cpuid_available");  // clear flag
+     __asm__(" push %ebx");
+     __asm__(" pushfd");
+     __asm__(" pop %eax");
+     __asm__(" mov %eax, %ebx");
+     __asm__(" xor 0x00200000, %eax");
+     __asm__(" push %eax");
+     __asm__(" popfd");
+     __asm__(" pushfd");
+     __asm__(" pop %eax");
+     __asm__(" xor %ebx, %eax");
+     __asm__(" je done");
+     __asm__(" mov $1, __cpuid_available");
+     goto done;  // just to shut the compiler up
+   done:
+     __asm__(" push %ebx");
+     __asm__(" popfd");
+     __asm__(" pop %ebx");
+#endif
+	HasCPUIDInstruction=!!cpuid_available;
+}
+```
+
+This code is attempting to check if the CPU has the `CPUID` instruction.
+
+My goto equivalent is:
+
+```c++
+#if _WIN32
+#include <intrin.h>
+#endif
+
+class CPUDetectClass {
+    public:
+    void Init_CPUID_Instruction();
+    private:
+    bool HasCPUIDInstruction = false;
+};
+
+void CPUDetectClass::Init_CPUID_Instruction()
+{
+	unsigned long cpuid_available=0;
+    unsigned int mask = 0x00200000;
+
+#ifdef _WIN32
+    unsigned int eflags = __readeflags();
+    unsigned int modifiedEflags = eflags ^ mask; // Flip the 21st bit (ID flag) of EFLAGS
+    __writeeflags(modifiedEflags);
+    
+    // Compare the original and new EFLAGS value at bit 21
+    cpuid_available = (unsigned long)((__readeflags() ^ eflags) & mask);
+#elif defined(_UNIX)
+    unsigned int eflags;
+    // GCC/Clang require inline assembly, sorry :(
+    // Save current EFLAGS to eflags
+    asm volatile("push\n\t"
+                 "pop %0"
+                 : "=r"(eflags));
+
+    unsigned int modifiedEflags = eflags ^ mask; // Flip the 21st bit (ID flag) of EFLAGS
+    // Write the modified EFLAGS
+    asm volatile("push %0\n\t"
+                 "popf"
+                 :
+                 : "r"(modifiedEflags));
+
+    // Read back the current EFLAGS to see if the ID flag changed
+    unsigned int newEflags;
+    asm volatile("pushf\n\t"
+                 "pop %0"
+                 : "=r"(newEflags));
+    
+    // Compare the original and new EFLAGS value at bit 21
+    cpuid_available = (unsigned long)((newEflags ^ eflags) & mask);
+#endif
+	HasCPUIDInstruction=!!cpuid_available;
```

Comment:
We can probably just avoid all of this and return true on any x86 CPU we are running on. Windows 2000/ME were the last OS to run on a CPU that might not have theCPUID instruction. We can certainly just return true for a 64bit CPU.